_studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_frame.cpp

```diff
@@ -0,0 +1,316 @@
+            std::default_delete<MPEG2Slice>()
+        );
+
+        slices.resize(0);
+        sliceCount = 0;
+        FreeReferenceFrames();
+
+    }
+
+    void MPEG2DecoderFrameInfo::AddSlice(MPEG2Slice* slice)
+    {
+        if (0 == sliceCount) // on the first slice
+        {
+            const auto pic    = slice->GetPicHeader();
+            const auto picExt = slice->GetPicExtHeader();
+            isField       =  picExt.picture_structure != FRM_PICTURE;
+            isBottomField = (picExt.picture_structure == BOTTOM_FLD_PICTURE);
+            frameType     = (FrameType)pic.picture_coding_type;
+        }
+
+        slices.push_back(slice);
+        sliceCount++;
+    }
+
+    void MPEG2DecoderFrameInfo::UpdateReferences(const DPBType & dpb)
+    {
+        const auto pic    = GetSlice(0)->GetPicHeader();
+        const auto picExt = GetSlice(0)->GetPicExtHeader();
+
+        if (MPEG2_I_PICTURE == pic.picture_coding_type)
+            return;
+
+        // 7.6.3.5 Prediction in P-pictures (for IP pair)
+        if (FRM_PICTURE != frame.pictureStructure)
+        {
+            uint8_t currFieldIndex = frame.GetNumberByParity(picExt.picture_structure == BOTTOM_FLD_PICTURE);
+            if ((1 == currFieldIndex) && (MPEG2_P_PICTURE == pic.picture_coding_type)) // current picture is P and the 2nd field
+            {
+                if (MPEG2_I_PICTURE == frame.GetAU(0)->GetSlice(0)->GetPicHeader().picture_coding_type) // the first field is I
+                {
+                    references.forward = &frame;
+                    references.forward->IncrementReference();
+                    return;
+                }
+            }
+        }
+
+        // 7.6.1 Frame/Field prediction
+        DPBType refFrames = dpb;
+        refFrames.remove_if(
+            [this](MPEG2DecoderFrame const * f)
+            {
+                return (f == &frame) || !f->IsRef();
+            }
+        );
+
+        // Find latest ref frame
+        auto i = std::max_element(std::begin(refFrames), std::end(refFrames),
+            [](MPEG2DecoderFrame const* f1, MPEG2DecoderFrame const* f2)
+            {
+                return f1->decOrder < f2->decOrder;
+            }
+        );
+
+        auto ref = (i == std::end(refFrames)) ? nullptr : *i;
+
+        // If no reference frame is found, use 'self' reference.
+        // This is to handle cases without leading I picture
+        if (!ref)
+        {
+            ref = &frame;
+        }
+
+        if (MPEG2_P_PICTURE == pic.picture_coding_type)
+        {
+            if (ref)
+            {
+                references.forward = ref;
+                references.forward->IncrementReference();
+            }
+        }
+        else // MPEG2_B_PICTURE
+        {
+            if (ref)
+            {
+                references.backward = ref;
+                references.backward->IncrementReference();
+            }
+
+            refFrames.remove(ref);
```

Comment:
If !ref were possible:
if !ref references.backward is unassigned. And future computations can be skipped?